### ISSUE TITLE
UseStatusCodePages does not handle 404 responses when hosted in IIS.

### ISSUE THREAD

**Issue by IsaacSee** [aspnet/Diagnostics#144]
Hello, 

When using **app.UseStatusCodePagesWithReExecute("/status-pages/{0}")** within a website hosted with IIS. If no one handle the request (ie. 404), the StatusCode middleware is not triggered, instead the 404 is sent by IIS default html status code pages.

What would be the way to ensure that if no route matched, our custom StatusCodePage does handle the response ?

**Comment by kichalla** [comment]
I remember me too hitting this issue (which does not happen in case of WebListener)

**Comment by Tratcher** [comment]
Put this at the end of your startup:
```
app.Run(context =>
{
  context.Response.StatusCode = 404;
  return Task.FromResult(0);
});
```

This terminates the request before it has a chance to fall back into IIS.

**Comment by IsaacSee** [comment]
Exactly what I was looking for, thanks @Tratcher.

**Comment by RehanSaeed** [comment]
This works but when I try to use ```application.UseStatusCodePagesWithReExecute("/error/{0}");``` in conjunction with this code I get 404 errors all the time. Is there some way of combining the two?

**Comment by Tratcher** [comment]
@RehanSaeed When you don't have the app.Run, what component serves your /error/{0} pages? If you're going to re-execute the request then something in your app after UseStatusCodePagesWithReExecute needs to handle the /errors/ requests.

Have you looked at the sample?
https://github.com/aspnet/Diagnostics/blob/dev/samples/StatusCodePagesSample/Startup.cs#L59

**Comment by dazinator** [comment]
I am struggling to get this to work, guidance greatly appreciated,

On startup in Configure() I am doing:

 `app.UseStatusCodePagesWithReExecute("/Home/Errors/{0}");`

I have a controller method to handle those requests:

```

 public class HomeController : Controller
    {    

        public IActionResult Errors(string errCode)
        {
            switch (errCode)
            {
                case "401":
                case "403":
                case "404":
                case "500":
                case "502":
                case "503":
                case "504":
                    return View($"~/Views/Shared/Errors/{errCode}.cshtml");
                default: return Error();
            }

        }

        public IActionResult Error()
        {
            return View("~/Views/Shared/Errors/General.cshtml");
        }       
      
    }

```

I don't have any app.Run() catch all method.

Now if I navigate to a non-existent URL, the `public IActionResult Errors(string errCode)` method in the controller is fired.. but the `errCode` is null.

If i look at the current request it has this path: /Home/Errors/404

My default route looks like: `{[default, {controller=Home}/{action=Index}/{id?}]}	

It looks to me like a simple model binding problem, in that the ccurrent request path is `/Home/Errors/404` but my controller method isn't having the last segment in the path bound as the errCode parameter. 

What is the best way to fix that?

**Comment by Tratcher** [comment]
Do you have to rename `errCode` to `id`?

**Comment by dazinator** [comment]
Yep! that fixed it!

Care to fill me in on why? Is it something to do with my default route? I admit I am not particularly up to speed with routing yet.

**Comment by Tratcher** [comment]
In your route you named the default parameter to bind to `id`. When model binding happened it couldn't find parameter `id`, so it didn't do anything. Imagine if your method took multiple parameters, how would model binding know what went where?

**Comment by dazinator** [comment]
Ahh ok I see. Yeah I was being a bit naive, thinking I could just append segments to my URL and they would magically be bound in order to my method parameters. My bad..

So in the end, I fixed this by fixing the routing, I had the option of `RouteAttribute`s

```
  
    [Route("Home")]
    public class HomeController : BaseController
    {

        [Route("Errors/{errorCode}")]
        public IActionResult Errors(string errorCode)
        {
            switch (errorCode)
```

Which worked. Or: 

```

     app.UseMvc(routes =>
            {

                routes.MapRoute("errors", "Home/Errors/{errorCode}", 
                    defaults: new { controller = "Home", action = "Errors" });

               // Add other routes / defualt route below
```

Thanks for being patient. Seems to work nicely now.

**Comment by ghooton** [comment]
Please note that UseStatusCodePagesWithReExecute must come before UseMvc. i.e.

            // Make sure to keep the same request that caused the error when redirecting for logging purposes.
            app.UseStatusCodePagesWithReExecute("/error", "?status={0}");

            app.UseMvc(routes =>
            {
                routes.MapRoute(
                    name: "default",
                    template: "{controller}/{action=Index}/{id?}");
            });

            app.Run(context =>
            {
                // Prevent IIS from returning its standard 404 page
                context.Response.StatusCode = 404;
                return Task.FromResult(0);
            });

**Comment by Tratcher** [comment]
@ghooton you don't need that Run statement, there's always a blank 404 at the end of the pipeline.
